cosima_cookbook/database.py

```diff
@@ -578,86 +578,93 @@ def build_index(
 
     indexed = 0
     for directory in [Path(d) for d in directories]:
-        # find all netCDF files in the experiment directory
-        files = find_files(directory, followsymlinks=followsymlinks)
-        expt = find_experiment(session, directory)
 
+        expt = find_experiment(session, directory)
         if expt is None:
             expt = NCExperiment(
                 experiment=str(directory.name), root_dir=str(directory.resolve())
             )
 
         print("Indexing experiment: {}".format(directory.name))
 
-        existing_files = set([f.ncfile for f in expt.ncfiles])
-
-        # Prune files that exist in the database but are not present
-        # on the filesystem
-        missing_files = existing_files - files
-        if len(missing_files) > 0:
-            # Make a list of NCFile objects to pass to prune_files
-            missing_files_objs = [f for f in expt.ncfiles if f.ncfile in missing_files]
-            prune_files(expt, session, missing_files_objs, delete=(prune == "delete"))
+        # find all netCDF files in the experiment directory
+        files = find_files(directory, followsymlinks=followsymlinks)
 
         if force:
-            # Delete entries that we're reindexing
-            for f in files & existing_files:
-                existing_ncfile = (
-                    session.query(NCFile)
-                    .join(NCExperiment)
-                    .filter(NCExperiment.experiment == str(directory.name))
-                    .filter(NCExperiment.root_dir == str(directory.resolve()))
-                    .filter(NCFile.ncfile == f)
-                    .one_or_none()
-                )
-                if existing_ncfile is not None:
-                    print("Re-indexing {}".format(f))
-                    session.delete(existing_ncfile)
-            session.flush()
+            # Prune all files to force re-indexing data
+            _prune_files(expt, session, {}, delete=True)
         else:
-            # Only pass files that are not already in DB
-            files = files - existing_files
+            # Prune files that exist in the database but are not present on disk
+            _prune_files(expt, session, files, delete=(prune == "delete"))
 
-        indexed += index_experiment(files, session, expt, client)
+        if len(files) > 0:
+            if len(expt.ncfiles) > 0:
```

Comment:
This probably works even if `ncfiles` is empty -- set generator over empty set is fine?

Reply:
I put in those tests because I was getting errors like this:
```python
E                   sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Can't resolve value for column experiments.id on object <NCExperiment at 0x14980124cdc0>; no value has been set for this column
E                   [SQL: SELECT ncfiles.ncfile AS ncfiles_ncfile 
E                   FROM ncfiles 
E                   WHERE ? = ncfiles.experiment_id AND ncfiles.ncfile NOT IN (?)]
E                   [parameters: [immutabledict({})]]

/home/502/aph502/.local/lib/python3.9/site-packages/sqlalchemy/orm/relationships.py:1772: StatementError
```
I just commented out that check and immediately got the same error in this test:

https://github.com/COSIMA/cosima-cookbook/blob/c23b089571ec19624ab91f6b1c2544c0f61e7d0b/test/test_indexing.py#L86

At this location.